File: A4_Krocket/main.py
import numpy as np
import matplotlib.pyplot as plt
import time

import concurrent.futures

def read_input(file_path):
    with open(file_path, 'r') as file:
        lines = file.readlines()
    
    n, r = map(int, lines[0].strip().split())
    gatePoints = []
    for line in lines[1:n+1]:
        x1, y1, x2, y2 = map(int, line.strip().split())
        gatePoints.append(((x1, y1), (x2, y2)))
    # Sort gates by their distance from the origin (0, 0)
    gates2 = gatePoints
    if (file_path == 'Runde_1/Beispiele/A4_Krocket/krocket3.txt'):
        gates2.sort(key=lambda gate: min(np.hypot(gate[0][0], gate[0][1]), np.hypot(gate[1][0], gate[1][1])))
    return gatePoints, r*2, gates2

def plot_gates(gates):
    for gate in gates:
        (x1, y1), (x2, y2) = gate
        plt.plot([x1, x2], [y1, y2])
    
    plt.xlabel('X')
    plt.ylabel('Y')
    plt.title('Gates Plot')
    plt.grid(True)

def plot_r_gates(lines):
    for n in range(len(lines)):
        (x1, y1), (x2, y2) = lines[n]
        if n == 0 or n == len(lines) - 1:
            plt.plot([x1, x2], [y1, y2], color='red', marker='o')
        else:
            plt.plot([x1, x2], [y1, y2], color='gray')
    
    plt.xlabel('X')
    plt.ylabel('Y')
    plt.title('Gates Plot')
    plt.grid(True)

def plot_faulty_gates(lines):
    for n in range(len(lines)):
        (x1, y1), (x2, y2) = lines[n]
        plt.plot([x1, x2], [y1, y2], color='orange', marker='o')



def find_intersection_vector(gates, width=1.0, setY = False):
    def extending_length(gates, width):
        length = 0
        (x1, y1), (x2, y2) = gates[0]
        length += np.hypot(x2 - x1, y2 - y1)
        return length/3
    def extend_vector(start_point, end_point, length):
        direction_vector = (end_point[0] - start_point[0], end_point[1] - start_point[1])
        vector_length = np.hypot(direction_vector[0], direction_vector[1])
        scale_factor = length / vector_length
        extended_start = (start_point[0] - direction_vector[0] * scale_factor, start_point[1] - direction_vector[1] * scale_factor)
        extended_end = (end_point[0] + direction_vector[0] * scale_factor, end_point[1] + direction_vector[1] * scale_factor)
        return extended_start, extended_end
    def count_crossed_gates(start_point, end_point, gates, width):
        def do_lines_intersect(p1, p2, q1, q2):
            def ccw(A, B, C):
                return (C[1] - A[1]) * (B[0] - A[0]) > (B[1] - A[1]) * (C[0] - A[0])
            return ccw(p1, q1, q2) != ccw(p2, q1, q2) and ccw(p1, p2, q1) != ccw(p1, p2, q2)

        def is_point_on_line(px, py, x1, y1, x2, y2):
            return (x2 - x1) * (py - y1) == (y2 - y1) * (px - x1) and min(x1, x2) <= px <= max(x1, x2) and min(y1, y2) <= py <= max(y1, y2)
        
        def move_point_by_half_width(point, direction_vector, width):
            length = np.hypot(direction_vector[0], direction_vector[1])
            perp_dx, perp_dy = -direction_vector[1] / length * width / 2, direction_vector[0] / length * width / 2
            return (point[0] + perp_dx, point[1] + perp_dy), (point[0] - perp_dx, point[1] - perp_dy)

        count = 0
        direction_vector = (end_point[0] - start_point[0], end_point[1] - start_point[1])
        start_point1, start_point2 = move_point_by_half_width(start_point, direction_vector, width)
        end_point1, end_point2 = move_point_by_half_width(end_point, direction_vector, width)
        start_point1, end_point1 = extend_vector(start_point1, end_point1, extending_length(gates, width))
        start_point2, end_point2 = extend_vector(start_point2, end_point2, extending_length(gates, width))
        faulty_gates = []
        for gate in gates:
            (x1, y1), (x2, y2) = gate
            if ((is_point_on_line(start_point1[0], start_point1[1], x1, y1, x2, y2) and
                 is_point_on_line(end_point1[0], end_point1[1], x1, y1, x2, y2) and
                 is_point_on_line(start_point2[0], start_point2[1], x1, y1, x2, y2) and
                 is_point_on_line(end_point2[0], end_point2[1], x1, y1, x2, y2)) or
                (do_lines_intersect(start_point1, end_point1, (x1, y1), (x2, y2)) and
                 do_lines_intersect(start_point2, end_point2, (x1, y1), (x2, y2)))):
                count += 1
            else:
                faulty_gates.append(gate)
        return count, [start_point1, end_point1, start_point2, end_point2], faulty_gates

    best_start_point = None
    best_end_point = None
    start_point1 = None
    end_point1 = None
    start_point2 = None
    end_point2 = None
    max_crossed_gates = 0
    faulty_gates = []

    def generate_points(gate, num_points):
        (x1, y1), (x2, y2) = gate
        return [(x1 + t * (x2 - x1), y1 + t * (y2 - y1)) for t in np.linspace(0, 1, num_points)]

    num_points = 15
    max_points = 150
    all_gates_crossed = False

    while not all_gates_crossed and num_points <= max_points:
        (x1_start, y1_start), (x2_start, y2_start) = gates[0]
        (x1_end, y1_end), (x2_end, y2_end) = gates[-1]

        start_points = generate_points(gates[0], num_points)
        # print(start_points)
        end_points = generate_points(gates[-1], num_points)

        for start_point in start_points:
            for end_point in end_points:
                crossed_gates, points, faulty_gate = count_crossed_gates(start_point, end_point, gates, width)
                if crossed_gates > max_crossed_gates:
                    max_crossed_gates = crossed_gates
                    best_start_point = start_point
                    best_end_point = end_point
                    start_point1, end_point1, start_point2, end_point2 = points
                    faulty_gates = faulty_gate
                    if max_crossed_gates == len(gates):
                        all_gates_crossed = True
                        break
            if all_gates_crossed:
                break

        if not all_gates_crossed:
            num_points += 20 # Increase the number of points to check

    # Create a vector from the best start point to the best end point
    intersection_vector = [best_start_point, best_end_point]
    print(f"Intersection Vector: Start {best_start_point}, End {best_end_point}")
    print(f"Vector Equation: r(t) = ({best_start_point[0]}, {best_start_point[1]}) + t * (({best_end_point[0] - best_start_point[0]}, {best_end_point[1] - best_start_point[1]}) for t in [0, 1]")
    print(f"Crossed Gates: {max_crossed_gates} out of {len(gates)}")
    direction_vector = (best_end_point[0] - best_start_point[0], best_end_point[1] - best_start_point[1])
    print(f"Start Point: {best_start_point}")
    print(f"Direction Vector: {direction_vector}")
    # return intersection_vector
    return extend_vector(best_start_point, best_end_point, extending_length(gates, width)), [[start_point1, end_point1], [start_point2, end_point2]], faulty_gates

def plot_intersection_vector(vector, width, gates):
    (x1, y1), (x2, y2) = vector
    plt.plot([x1, x2], [y1, y2], marker='x', color='red', linestyle='--', label='Intersection Vector')

    dx, dy = x2 - x1, y2 - y1
    length = np.hypot(dx, dy)
    perp_dx, perp_dy = -dy / length * width*0.5, dx / length * width*0.5

    # Create the polygon for the width area
    polygon_points = [
        (x2 + perp_dx, y2 + perp_dy),
        (x2 - perp_dx, y2 - perp_dy),
        (x1 - perp_dx, y1 - perp_dy),
        (x1 + perp_dx, y1 + perp_dy),
    ]
    polygon = plt.Polygon(polygon_points, color='blue', alpha=0.3, label='Width Area')
    plt.gca().add_patch(polygon)

    plt.legend()

def plot_paralell_vector(vector):
    (x1, y1), (x2, y2) = vector
    plt.plot([x1, x2], [y1, y2], marker='x', color='green', linestyle='--', label='Intersection Vector')
    plt.legend()

def main():
    global_start_time = time.time()
    file_paths = [
        'Runde_1/Beispiele/A4_Krocket/krocket1.txt',
        'Runde_1/Beispiele/A4_Krocket/krocket2.txt',
        'Runde_1/Beispiele/A4_Krocket/krocket3.txt',
        'Runde_1/Beispiele/A4_Krocket/krocket4.txt',
        'Runde_1/Beispiele/A4_Krocket/krocket5.txt'
    ]

    file_paths_num = {
        'Runde_1/Beispiele/A4_Krocket/krocket1.txt': 1,
        'Runde_1/Beispiele/A4_Krocket/krocket2.txt': 2,
        'Runde_1/Beispiele/A4_Krocket/krocket3.txt': 3,
        'Runde_1/Beispiele/A4_Krocket/krocket4.txt': 4,
        'Runde_1/Beispiele/A4_Krocket/krocket5.txt': 5
    }
    
    for file_path in file_paths:
        print(f"Processing file: {file_path}")
        gatePoints, width, gates2 = read_input(file_path)
        plot_r_gates(gates2)
        print(f"Width: {width}")
        start_time = time.time()
        intersection_vector, paralell_vectors, faulty_gates = find_intersection_vector(gates2, width)
        end_time = time.time()
        
        print(f"Time taken to find the intersection vector: {end_time - start_time} seconds")
        
        plot_intersection_vector(intersection_vector, width, gates2)
        plot_paralell_vector(paralell_vectors[0])
        plot_paralell_vector(paralell_vectors[1])
        plot_faulty_gates(faulty_gates)
        plt.savefig(f"Runde_1/Beispiele/A4_Krocket/krocket_plot_{file_paths_num.get(file_path)}.svg", format='svg')
        plt.clf()
    global_end_time = time.time()
    print(f"Total time taken: {global_end_time - global_start_time} seconds")

if __name__ == '__main__':
    main()

File: A4_Krocket/src/main.rs
use std::fs::File;
use std::io::Write;
use std::io::{self, BufRead};
use std::time::Instant;

fn read_input(file_path: &str) -> (Vec<((i32, i32), (i32, i32))>, f64, Vec<((i32, i32), (i32, i32))>) {
    let file = File::open(file_path).expect("Unable to open file");
    let reader = io::BufReader::new(file);
    let mut lines = reader.lines().map(|l| l.unwrap());

    let first_line = lines.next().unwrap();
    let (n, r) = {
        let mut parts = first_line.split_whitespace();
        let n = parts.next().unwrap().parse::<i32>().unwrap();
        let r = parts.next().unwrap().parse::<f64>().unwrap();
        (n, r)
    };

    let mut gate_points = Vec::new();
    for _ in 0..n {
        let line = lines.next().unwrap();
        let mut parts = line.split_whitespace();
        let x1 = parts.next().unwrap().parse::<i32>().unwrap();
        let y1 = parts.next().unwrap().parse::<i32>().unwrap();
        let x2 = parts.next().unwrap().parse::<i32>().unwrap();
        let y2 = parts.next().unwrap().parse::<i32>().unwrap();
        gate_points.push(((x1, y1), (x2, y2)));
    }

    let mut gates2 = gate_points.clone();
    if file_path == "./A4_Krocket/krocket3.txt" {
        gates2.sort_by(|a, b| {
            let dist_a = ((a.0 .0 as f64).hypot(a.0 .1 as f64)).min((a.1 .0 as f64).hypot(a.1 .1 as f64));
            let dist_b = ((b.0 .0 as f64).hypot(b.0 .1 as f64)).min((b.1 .0 as f64).hypot(b.1 .1 as f64));
            dist_a.partial_cmp(&dist_b).unwrap()
        });
    }

    (gate_points, r*2.0, gates2)
}

extern crate plotters;
use plotters::prelude::*;
use plotters::style::WHITE;
use plotters::style::IntoFont;
use plotters::chart::ChartBuilder;
use plotters::series::LineSeries;
use plotters::backend::BitMapBackend;

fn plot_gate_points(gate_points: Vec<((i32, i32), (i32, i32))>, intersection_vector: ((f64, f64), (f64, f64)), parallel_vectors: Vec<((f64, f64), (f64, f64))>, file_number: &str) -> Result<(), Box<dyn std::error::Error>> {
    let file_name = format!("plot_{}.svg", file_number);
    let root = SVGBackend::new(&file_name, (1280, 960)).into_drawing_area();
    root.fill(&WHITE)?;

    let mut chart = ChartBuilder::on(&root)
        .caption("Gate Points and Intersection Vector", ("sans-serif", 50).into_font())
        .margin(10)
        .x_label_area_size(30)
        .y_label_area_size(30)
        .build_cartesian_2d(
            gate_points.iter().flat_map(|&((x1, _), (x2, _))| vec![x1, x2]).min().unwrap()..gate_points.iter().flat_map(|&((x1, _), (x2, _))| vec![x1, x2]).max().unwrap(),
            gate_points.iter().flat_map(|&((_, y1), (_, y2))| vec![y1, y2]).min().unwrap()..gate_points.iter().flat_map(|&((_, y1), (_, y2))| vec![y1, y2]).max().unwrap()
        )?;
    chart.configure_mesh().draw()?;

    for gate in &gate_points {
        chart.draw_series(LineSeries::new(
            vec![(gate.0 .0, gate.0 .1), (gate.1 .0, gate.1 .1)],
            &plotters::style::RED,
        ))?;
    }

    chart.draw_series(LineSeries::new(
        vec![(intersection_vector.0 .0 as i32, intersection_vector.0 .1 as i32), (intersection_vector.1 .0 as i32, intersection_vector.1 .1 as i32)],
        &plotters::style::BLUE,
    ))?;

    for vector in &parallel_vectors {
        chart.draw_series(LineSeries::new(
            vec![(vector.0 .0 as i32, vector.0 .1 as i32), (vector.1 .0 as i32, vector.1 .1 as i32)],
            &plotters::style::BLUE,
        ))?;
    }

    root.present()?;
    Ok(())
}

fn find_intersection_vector(gates: &Vec<((i32, i32), (i32, i32))>, width: f64) -> (((f64, f64), (f64, f64)), Vec<((f64, f64), (f64, f64))>, Vec<((i32, i32), (i32, i32))>) {
    fn extending_length(gates: &Vec<((i32, i32), (i32, i32))>, _width: f64) -> f64 {
        gates.iter().map(|&((x1, y1), (x2, y2))| {
            (((x2 - x1).pow(2) as f64 + (y2 - y1).pow(2) as f64).sqrt() / 3.0) as f64
        }).sum()
    }

    fn extend_vector(start_point: (f64, f64), end_point: (f64, f64), length: f64) -> ((f64, f64), (f64, f64)) {
        let direction_vector = (end_point.0 - start_point.0, end_point.1 - start_point.1);
        let vector_length = (direction_vector.0.powi(2) + direction_vector.1.powi(2)).sqrt();
        let scale_factor = length / vector_length;
        let extended_start = (start_point.0 - direction_vector.0 * scale_factor.min(1.0), start_point.1 - direction_vector.1 * scale_factor.min(1.0));
        let extended_end = (end_point.0 + direction_vector.0 * scale_factor.min(1.0), end_point.1 + direction_vector.1 * scale_factor.min(1.0));
        (extended_start, extended_end)
    }

    fn count_crossed_gates(start_point: (f64, f64), end_point: (f64, f64), gates: &Vec<((i32, i32), (i32, i32))>, width: f64) -> (usize, Vec<((f64, f64), (f64, f64))>, Vec<((i32, i32), (i32, i32))>) {
        fn ccw(a: (f64, f64), b: (f64, f64), c: (i32, i32)) -> bool {
            (c.1 as f64 - a.1) * (b.0 - a.0) > (b.1 as f64 - a.1) * (c.0 as f64 - a.0)
        }

        fn do_lines_intersect(p1: (f64, f64), p2: (f64, f64), q1: (i32, i32), q2: (i32, i32)) -> bool {
            ccw(p1, (q1.0 as f64, q1.1 as f64), (q2.0, q2.1)) != ccw(p2, (q1.0 as f64, q1.1 as f64), (q2.0, q2.1)) && ccw(p1, p2, (q1.0, q1.1)) != ccw(p1, p2, (q2.0, q2.1))
        }

        let mut count = 0;

        fn is_point_on_line(px: f64, py: f64, x1: i32, y1: i32, x2: i32, y2: i32) -> bool {
            (x2 - x1) as f64 * (py - y1 as f64) == (y2 - y1) as f64 * (px - x1 as f64) && x1.min(x2) as f64 <= px && px <= x1.max(x2) as f64 && y1.min(y2) as f64 <= py && py <= y1.max(y2) as f64
        }

        fn move_point_by_half_width(point: (f64, f64), direction_vector: (f64, f64), width: f64) -> ((f64, f64), (f64, f64)) {
            let length = (direction_vector.0.powi(2) + direction_vector.1.powi(2)).sqrt();
            let perp_dx = -direction_vector.1 / length * width / 2.0;
            let perp_dy = direction_vector.0 / length * width / 2.0;
            ((point.0 + perp_dx, point.1 + perp_dy), (point.0 - perp_dx, point.1 - perp_dy))
        }

        let direction_vector = (end_point.0 - start_point.0, end_point.1 - start_point.1);
        let (start_point1, start_point2) = move_point_by_half_width(start_point, direction_vector, width);
        let (end_point1, end_point2) = move_point_by_half_width(end_point, direction_vector, width);
        let (start_point1_ext, end_point1_ext) = extend_vector(start_point1, end_point1, extending_length(gates, width));
        let (start_point2_ext, end_point2_ext) = extend_vector(start_point2, end_point2, extending_length(gates, width));
        let mut faulty_gates = Vec::new();

        for gate in gates {
            let ((x1, y1), (x2, y2)) = *gate;
            if (is_point_on_line(start_point1_ext.0, start_point1_ext.1, x1, y1, x2, y2) &&
                is_point_on_line(end_point1_ext.0, end_point1_ext.1, x1, y1, x2, y2) &&
                is_point_on_line(start_point2_ext.0, start_point2_ext.1, x1, y1, x2, y2) &&
                is_point_on_line(end_point2_ext.0, end_point2_ext.1, x1, y1, x2, y2)) ||
                (do_lines_intersect(start_point1_ext, end_point1_ext, (x1, y1), (x2, y2)) &&
                 do_lines_intersect(start_point2_ext, end_point2_ext, (x1, y1), (x2, y2))) {
                count += 1;
            } else {
                faulty_gates.push(*gate);
            }
        }
        (count, vec![(start_point1, end_point1), (start_point2, end_point2)], faulty_gates)
    }

    fn generate_points(gate: ((i32, i32), (i32, i32)), num_points: usize) -> Vec<(f64, f64)> {
        let ((x1, y1), (x2, y2)) = gate;
        (0..num_points).map(|t| {
            let t = t as f64 / (num_points - 1) as f64;
            (x1 as f64 + t * (x2 as f64 - x1 as f64), y1 as f64 + t * (y2 as f64 - y1 as f64))
        }).collect()
    }

    let mut best_start_point = None;
    let mut best_end_point = None;
    let mut start_point1 = None;
    let mut end_point1 = None;
    let mut start_point2 = None;
    let mut end_point2 = None;
    let mut max_crossed_gates = 0;
    let mut faulty_gates = Vec::new();

    let mut num_points = 15;
    let max_points = 150;
    let mut all_gates_crossed = false;

    while !all_gates_crossed && num_points <= max_points {
        let start_points = generate_points(gates[0], num_points);
        let end_points = generate_points(gates[gates.len() - 1], num_points);

        for start_point in &start_points {
            for end_point in &end_points {
                let (crossed_gates, points, _faulty_gate) = count_crossed_gates(*start_point, *end_point, gates, width);
                if crossed_gates > max_crossed_gates {
                    max_crossed_gates = crossed_gates;
                    best_start_point = Some(*start_point);
                    best_end_point = Some(*end_point);
                    start_point1 = Some(points[0]);
                    start_point2 = Some(points[1]);
                    end_point1 = Some(points[0]);
                    end_point2 = Some(points[1]);
                    faulty_gates = _faulty_gate;
                    if max_crossed_gates == gates.len() {
                        all_gates_crossed = true;
                        break;
                    }
                }
            }
            if all_gates_crossed {
                break;
            }
        }

        if !all_gates_crossed {
            num_points += 20;
        }
    }

    let best_start_point = best_start_point.unwrap();
    let best_end_point = best_end_point.unwrap();
    // let intersection_vector = extend_vector(best_start_point, best_end_point, extending_length(gates, width));
    let intersection_vector = (best_start_point, best_end_point);
    println!("Intersection Vector: Start {:?}, End {:?}", best_start_point, best_end_point);
    println!("Vector Equation: r(t) = ({}, {}) + t * (({}, {})) for t in [0, 1]", best_start_point.0, best_start_point.1, best_end_point.0 - best_start_point.0, best_end_point.1 - best_start_point.1);
    println!("Crossed Gates: {} out of {}", gates.len() - faulty_gates.len(), gates.len());
    let direction_vector = (best_end_point.0 - best_start_point.0, best_end_point.1 - best_start_point.1);
    println!("Start Point: {:?}", best_start_point);
    println!("Direction Vector: {:?}", direction_vector);
    (intersection_vector, vec![start_point1.unwrap(), end_point1.unwrap(), start_point2.unwrap(), end_point2.unwrap()], faulty_gates)
}

fn main() {
    let global_start_time = Instant::now();
    let file_paths = [
        "./A4_Krocket/krocket1.txt",
        "./A4_Krocket/krocket2.txt",
        "./A4_Krocket/krocket3.txt",
        "./A4_Krocket/krocket4.txt",
        "./A4_Krocket/krocket5.txt",
    ];
    let mut file_dict = std::collections::HashMap::new();
    for (i, file_path) in file_paths.iter().enumerate() {
        file_dict.insert(*file_path, (i + 1).to_string());
    }

    for file_path in &file_paths {
        println!("Processing file: {}", file_path);
        let (gate_points, width, gates2) = read_input(file_path);

        let start_time = Instant::now();
        let (intersection_vector, _parallel_vectors, faulty_gates) = find_intersection_vector(&gates2, width);
        let gates_crossed = gates2.len() - faulty_gates.len();
        println!("Number of gates crossed: {}", gates_crossed);
        let end_time = Instant::now();
        println!("Intersection Vector: Start {:?}, End {:?}", intersection_vector.0, intersection_vector.1);
        println!("Time taken to find the intersection vector: {:?}", end_time.duration_since(start_time));
        // Write the data to a file
        let output_file_name = format!("output_{}.txt", file_dict.get(file_path).unwrap());
        let mut output_file = File::create(&output_file_name).expect("Unable to create output file");

        writeln!(output_file, "Intersection Vector: Start {:?}, End {:?}", intersection_vector.0, intersection_vector.1).expect("Unable to write data");
        writeln!(output_file, "Vector Equation: r(t) = ({}, {}) + t * (({}, {})) for t in [0, 1]", intersection_vector.0 .0, intersection_vector.0 .1, intersection_vector.1 .0 - intersection_vector.0 .0, intersection_vector.1 .1 - intersection_vector.0 .1).expect("Unable to write data");
        writeln!(output_file, "Crossed Gates: {} out of {}", gates2.len() - faulty_gates.len(), gates2.len()).expect("Unable to write data");
        writeln!(output_file, "Start Point: {:?}", intersection_vector.0).expect("Unable to write data");
        writeln!(output_file, "Direction Vector: {:?}", (intersection_vector.1 .0 - intersection_vector.0 .0, intersection_vector.1 .1 - intersection_vector.0 .1)).expect("Unable to write data");
        
        // Call the plotting function
        if let Err(e) = plot_gate_points(gates2.clone(), intersection_vector, _parallel_vectors.clone(), file_dict.get(file_path).unwrap()) {
            eprintln!("Error plotting gate points: {}", e);
        }
    }
    let global_end_time = Instant::now();
    println!("Total time taken: {:?}", global_end_time.duration_since(global_start_time));

    // Write the total time taken to a file
    let mut total_time_file = File::create("total_time.txt").expect("Unable to create total time file");
    writeln!(total_time_file, "Total time taken: {:?}", global_end_time.duration_since(global_start_time)).expect("Unable to write total time");
}